### DISCUSSION TITLE
Performance of BlockNonlinearForm compared to NonlinearForm

### DISCUSSION THREAD

**Discussion by thelfer** [mfem/mfem#2229] [General]
I just wanted to know if anyone already tested if using a `BlockNonlinearForm` has a performance hit compared to a `NonlinearForm` for problems with a unique unknowns.

The context of this question is that I am writting an application on top of `MFEM` for (generalised) non linear solid mechanics. So I will solve problem with only one unknowns (the displacement), and other with several unknowns (mechanics strongly coupled with heat transfer for example). It would be nice to avoid code duplication, so I wondered if building all my code around  `BlockNonlinearForm` from the start would be a good idea, whatever the problem treated.

Thanks for any advice !

**Comment by jandrej** [comment]
There should be no performance overhead using `BlockNonlinearForm`. We have no formal benchmarks to present though.

**Reply by thelfer** [reply]
@jandrej Thanks for the anwser.

As a corollary, is there any advantage of using  `NonlinearForm` rather than `BlockNonlinearForm`? Sorry if the question is a bit rude

**Reply by jandrej** [reply]
`BlockNonlinearForm` evolved out of the necessity to have a more convenient interface for multi-variable problems. If you only have a problem with one variable you could use `NonlinearForm` to avoid the `Array` notation. I don't see any other major advantage.

**Reply by thelfer** [reply]
@jandrej Thanks again. I'll thus change my app to use `BlockNonlinearForm` for each case.
